include/tvm/layout.h

```diff
@@ -0,0 +1,435 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file tvm/layout.h
+ * \brief Layout expression.
+ *
+ *  This file is adapted from its nnvm counterpart and will keep involving
+ *  to the new layout system
+ *
+ *  The layout is composed of upper cases, lower cases and numbers,
+ *  where upper case indicates a primal axis and
+ *  the corresponding lower case with factor size indicates the subordinate axis.
+ *  For example, NCHW16c can describe a 5-D tensor of
+ *  [batch_size, channel, height, width, channel_block].
+ *  Here subordinate axis channel_block=16 is the factor size of the primal axis C (channel).
+ */
+#ifndef TVM_LAYOUT_H_
+#define TVM_LAYOUT_H_
+
+#include <tvm/base.h>
+#include <tvm/expr.h>
+
+#include <string>
+#include <sstream>
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+#include "ir_operator.h"
+
+namespace tvm {
+
+class LayoutAxis {
+ public:
+  // single axis definitions
```

Comment:
Consider a change to array? move static construction to src, unless we need to manipulate the variable name explicitly. LayoutAxis.Get('A') seems to be as convenient